File: ImperatorToCK3/Source/CommonUtilities/SimpleField.cpp
#include "SimpleField.h"



SimpleField::SimpleField(std::optional<std::string> initialValue) : initialValue(std::move(initialValue)) {}

std::optional<std::string> SimpleField::getValue(const date& date) const {
	if (const auto& lowerBound = valueHistory.lower_bound(date); lowerBound != valueHistory.end()) {
		return lowerBound->second;
	}
	return initialValue;
}

File: ImperatorToCK3/CommonUtils/SimpleField.cs
using System.Collections.Generic;
using System.Linq;
using commonItems;

namespace ImperatorToCK3.CommonUtils {
	public class SimpleField {
		public SortedDictionary<Date, string> ValueHistory { get; private set; } = new();
		public string? InitialValue { private get; set; }

		public SimpleField(string? initialValue) {
			InitialValue = initialValue;
		}
		public string? GetValue(Date date) {
			var pairWithLastEarlierOrSameDate = ValueHistory.TakeWhile(d => d.Key <= date);
			if (pairWithLastEarlierOrSameDate.Any()) {
				return pairWithLastEarlierOrSameDate.Last().Value;
			}
			return InitialValue;
		}
		public void AddValueToHistory(string value, Date date) {
			ValueHistory[date] = value;
		}
	}
}
